src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs

```diff
@@ -269,7 +269,7 @@ internal static string ReadToken(string data, ref int offset, StringBuilder? bui
 
             if (start == offset)
             {
-                throw new FormatException(SR.Format(SR.MailHeaderFieldInvalidCharacter, data[offset]));
+                throw new FormatException(SR.ContentTypeInvalid);
```

Comment:
What if we return either empty or `null` string in this case? ... Rather than embed semantic knowledge into token parser?

One usage can deal with it:
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Net.Mail/src/System/Net/Mime/ContentType.cs#L265-L270

Another usage cannot (I assume it will throw `IndexOutOfRangeException`) - maybe we should fix that rather than proposed change?
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Net.Mail/src/System/Net/Mime/ContentDisposition.cs#L302-L309

Reply:
Both ReadParameterAttribute and ReadToken are being invoked in various places. If we stop throwing an exception here, it might not be a trivial change and I'm not sure if the tests are good enough and cover every edge case. I am not too familiar with all the parsing rules so I don't want to unintentionally introduce a breaking change. 

But I agree that token parser doesn't need to know about content type semantics. How about instead of "ContentTypeInvalid" we throw a slightly more generic exception - like the existing error "InvalidHeaderName" (An invalid character was found in header name)?

Reply:
If we throw different exception, then we would have to just change the exception into another one in the caller IMO.

I didn't realize `ReadToken` is used directly, that makes things more complicated -- I am hesitant to start throwing different exception from the method without fully understanding all the usages and how it propagates from there. I think a proper fix should review all the usages of the parser methods and do the right thing IMO. Maybe adding comments and defining proper contract would be useful?

Reply:
> If we throw different exception, then we would have to just change the exception into another one in the caller IMO.
> 

I found that in all callers ([except this one](https://github.com/dotnet/runtime/blob/fe41d3c762695cfeb364a8a0b61e2e376d22da5a/src/libraries/System.Net.Mail/src/System/Net/Mime/ContentType.cs#L113-L118)), we are already changing the exception in the caller.


I agree that we need a good contract definition for the parser. However, the scope of this PR is to fix the `IndexOutOfRangeException` thrown specifically because we are accessing `data[offset]` while throwing `FormatException` [here](https://github.com/dotnet/runtime/blob/fe41d3c762695cfeb364a8a0b61e2e376d22da5a/src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs#L272).

If changing the exception message doesn't look good, another option is to change the condition which throws this exception:
[`if (start == offset)`](https://github.com/dotnet/runtime/blob/fe41d3c762695cfeb364a8a0b61e2e376d22da5a/src/libraries/Common/src/System/Net/Mail/MailBnfHelper.cs#L270) can be changed to `if (start == offset && offset < data.Length)`

For more in depth changes to this code, I propose we create a new work item for a proper fix.

Reply:
Hi @karelz , do these changes look good?